### ISSUE TITLE
Bad record type returned for selectFrom(T) when T contains LEFT SEMI JOIN or LEFT ANTI JOIN

### ISSUE THREAD

**Issue by bendem** [jOOQ/jOOQ#10183]
### Expected behavior and actual behavior:

I was trying to get a type safe object from a join query so I ended up reading #6503.

After applying the advice there, I came up with this query:

```java
public Result<MovementsRecord> missingSubscription(YearMonth month) {
    // in case you wonder, this is https://blog.jooq.org/2015/10/13/semi-join-and-anti-join-should-have-its-own-syntax-in-sql/
    return dsl
        .selectFrom(
            MOVEMENTS
                .leftSemiJoin(MOVEMENT_CATEGORIES).on(
                    MOVEMENTS.CATEGORY_ID.eq(MOVEMENT_CATEGORIES.ID)
                        .and(MOVEMENT_CATEGORIES.NAME.eq("Cotisations"))
                )
                .leftAntiJoin(SUBSCRIPTIONS).onKey(Keys.SUBSCRIPTIONS__SUBSCRIPTIONS_MOVEMENT_ID_FKEY)
        )
        .where(MOVEMENTS.ENTRY_DATE.between(month.atDay(1), month.atEndOfMonth()))
        .fetch();
}
```

Through generics magic, this compiles and runs, except when trying to access the returned object, I'm getting a lovely

```
java.lang.ClassCastException: class org.jooq.impl.RecordImpl 
cannot be cast to class be.lghs.accounting.model.tables.records.MovementsRecord
```

from this code
```java
Result<MovementsRecord> movements = movementRepository.missingSubscription(month);
System.out.println(movements.get(0).getId());
```

### Versions:

- jOOQ: 3.12.4
- Java: 11
- Database (include vendor): postgresql 12
- OS: fedora 31 (not that it matters)
- JDBC Driver (include name if inofficial driver): org.postgresql:postgresql:42.2.12

**Comment by bendem** [comment]
Note that the goal is to get a `Result<MovementsRecord>` from a query that does join, I'd be happy to do it some other way (I expected `select(MOVEMENTS.asterisk())` to do the trick but it seems more complex than that).

Using `RecordX` is not great since you lose the field names and have to rely on type instead.

**Comment by bendem** [comment]
For reference, this is what I tried, I desugared the semi/anti joins until this is looked at.
https://github.com/LgHS/accounting/blob/f5445b994c682f1f210a6514656862bb925dd924/src/main/java/be/lghs/accounting/repositories/MovementRepository.java#L194-L205

**Comment by lukaseder** [comment]
Thanks for your report. I can reproduce this when combining `LEFT SEMI JOIN` with `LEFT ANTI JOIN`. Will investigate

**Comment by lukaseder** [comment]
Hmm, actually, no combination is needed, it already happens with a single `LEFT SEMI JOIN`. Our integration tests bypassed the record assignment, and as such, never ran into this issue.

**Comment by lukaseder** [comment]
Fixed for jOOQ 3.14. Backport scheduled for 3.13.3 (#10185) and 3.12.5 (#10186)

**Comment by bendem** [comment]
Awesome, thanks for the quick fix!
